src/azul/drs.py

```diff
@@ -74,10 +87,164 @@ class Access:
     headers: Optional[Mapping[str, str]] = None
 
 
+@attr.s()
+class DRSURI(metaclass=ABCMeta):
+
+    @property
+    @abstractmethod
+    def scheme(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def object_id(self) -> str:
+        raise NotImplementedError
+
+    @classmethod
+    def parse(cls, drs_uri: str) -> 'DRSURI':
+        try:
+            drs_uri = RegularDRSURI.parse(drs_uri)
+        except ValueError as e:
+            if len(e.args) == 1 and isinstance(e.args[0], str) and e.args[0].startswith('Invalid port '):
+                drs_uri = CompactDRSURI.parse(drs_uri)
+            else:
+                raise
+        return drs_uri
+
+    def __attrs_post_init__(self):
+        assert self.scheme == 'drs', self
+
+    def to_url(self, client: 'DRSClient', access_id: Optional[str] = None) -> str:
+        """
+        Translate the DRS URI into a DRS URL. All query params included in the DRS
+        URI (eg '{drs_uri}?version=123') will be carried over to the DRS URL.
+        Compact, identifier-based URIs (drs://[provider_code/]namespace:accession)
+        are expanded to hostname-based DRS URIs (drs://<hostname>/<id>).
```

Comment:
This comment still puzzles me. Conversions from DRS URI to DRS URI is not defined in the DRS spec. Only the conversion from URI to URL is.

I think you are being misled by a previous attempt to handle/prevent compact identifiers (part of #2150).